### PR TITLE
Add `bits shl` and `bits shr` command

### PR DESCRIPTION
**Pull request by hustcer**
# Description

Add `bits shl` and `bits shr` command, need carefully review:
```shell
2 | bits shl 7
---
0

2 | bits shl 7 -n 2
---
256

0x7F | bits shl 1 -s
---
-2

[5 3 2] | bits shl 2
---
 0   20
 1   12
 2    8

8 | bits shr 2
---
2

[15 35 2] | bits shr 2
---
 0   3
 1   8
 2   0
```

# Tests

Make sure you've done the following:

- [x] Add tests that cover your changes, either in the command examples, the crate/tests folder, or in the /tests folder.
- [ ] Try to think about corner cases and various ways how your changes could break. Cover them with tests.
- [ ] If adding tests is not possible, please document in the PR body a minimal example with steps on how to reproduce so one can verify your change works.

Make sure you've run and fixed any issues with these commands:

- [x] `cargo fmt --all -- --check` to check standard code formatting (`cargo fmt --all` applies these changes)
- [x] `cargo clippy --workspace --features=extra -- -D warnings -D clippy::unwrap_used -A clippy::needless_collect` to check that you're using the standard code style
- [x] `cargo test --workspace --features=extra` to check that all the tests pass

### CODE REVIEW

**Comment by fdncred** [comment]
nice work @hustcer. I'll leave it open to see if someone else wants to give it a review.

**Comment by hustcer** [comment]
@sholderbach @WindSoilder Any thoughts?

**Comment by WindSoilder** [comment]
Looks good to me :-D

File: crates/nu-command/src/bits/shift_left.rs [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+            let shift_result = i64::try_from(val);
+            match shift_result {
+                Ok(val) => Value::Int { val, span },
+                Err(_) => Value::Error {
+                    error: ShellError::GenericError(
+                        "Shift left result beyond the range of 64 bit signed number".to_string(),
+                        format!(
+                            "{} of the specified number of bytes shift left {} bits exceed limit",
+                            val, bits
+                        ),
```
Comment (sholderbach) [inline crates/nu-command/src/bits/shift_left.rs:125]: I think overflowing might be a valid use case for shift operations, but warning of the overflow might well be the safer default.

File: crates/nu-command/src/bits/shift_right.rs [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+            .required("bits", SyntaxShape::Int, "number of bits to shift right")
+            .switch(
+                "signed",
+                "always treat input number as a signed number",
+                Some('s'),
+            )
+            .named(
+                "number-bytes",
+                SyntaxShape::String,
+                "the size of number in bytes, it can be 1, 2, 4, 8, auto, default value `auto`",
```
Comment (sholderbach) [inline crates/nu-command/src/bits/shift_right.rs:30]:
```suggestion
                "the word size in number of bytes, it can be 1, 2, 4, 8, auto, default value `auto`",
```

**Review by sholderbach — COMMENTED** [review]
Small nits about the error messages. 

I assume you will do the `bits rotate` next? :P

File: crates/nu-command/src/bits/shift_left.rs (hunk 1 above)
Comment (hustcer) [reply crates/nu-command/src/bits/shift_left.rs:125]: So what should I do? Can you explain with more detail?

**Comment by hustcer** [comment]
> Small nits about the error messages. 
> 
> 
> 
> I assume you will do the `bits rotate` next? :P

Yeah, in one or two days

File: crates/nu-command/src/bits/shift_left.rs (hunk 1 above)
Comment (hustcer) [reply crates/nu-command/src/bits/shift_left.rs:125]: Yeah, that's it

Comment (sholderbach) [reply crates/nu-command/src/bits/shift_left.rs:125]: Yeah, without taking on either an unsigned type, writing it as `Value::Bytes` or getting back to bigints that will not be representable. At some point somebody rolling their own crypto in nushell will stumble upon this and complain. But I think the PR is ready to land for the general use case

Comment (hustcer) [reply crates/nu-command/src/bits/shift_left.rs:125]: Thanks, let's land it

**Comment by hustcer** [comment]
@sholderbach That's because you didn't specify `--number-bytes`, the default value will be auto, then it actually run `1 | bits shl 62 -n 1`, you can try '1 | bits shl 62 -n 8'

**Comment by hustcer** [comment]
The default value is `8` now

**Comment by sholderbach** [comment]
> The default value is `8` now

That is probably the least surprising default. 

The internal type inference could have its lower bound not only based on the size required for the operand but also bounded by the demanded shift. But that is going beyond dotting the "i"s

**Comment by fdncred** [comment]
thanks. let's land this.
